### PR TITLE
Single strategy Locked withdrawal Escrow

### PR DESCRIPTION
**Pull request by maruf0011**
Epoch based locked withdrawal contract for single strategy vault. 
- [x] Contract.
- [x] Base test with single user.
- [x] Test with multi user withdrawal.

### CODE REVIEW

File: src/vaults/SingleStrategyWithdrawal.sol [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    /**
+     * @notice resolve the locked shares for current epoch
+     * @dev This function will be triggered after closing a position
+     * @dev will check for available shares to burn
+     * @dev after resolving vault will send the assets to escrow and burn the share
+     */
+    function resolveDebtShares() external onlyVault {
+        uint256 shares = vault.balanceOf(address(this));
```
Comment (larryob) [inline src/vaults/SingleStrategyWithdrawal.sol:89]: We resolve shares 1 epoch at a time. So we should only be attempting to burn shares for the epoch that was just closed.

File: src/vaults/SingleStrategyWithdrawal.sol [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        lastResolvedUTCTime = block.timestamp;
+    }
+
+    /**
+     * @notice Redeem withdrawal request
+     * @param user address
+     * @param epoch withdrawal request epoch
+     * @return received assets
+     */
+    function redeem(address user, uint256 epoch) external returns (uint256) {
```
Comment (larryob) [inline src/vaults/SingleStrategyWithdrawal.sol:113]: A user should only be able to redeem their own shares.

File: src/vaults/SingleStrategyWithdrawal.sol [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        // Should be a resolved epoch
+        require(canWithdraw(epoch), "SSWE: epoch not resolved.");
+
+        // total assets for user
+        uint256 assets = epochAsset[epoch].mulDivDown(userDebtShare[epoch][user], epochDebt[epoch]);
+
+        // check for asset balance
+        require(assets <= asset.balanceOf(address(this)), "SSWE: Not enough asset.");
```
Comment (larryob) [inline src/vaults/SingleStrategyWithdrawal.sol:121]: The transfer will fail if there's not enough assets.

**Review by larryob — COMMENTED** [review]
Looking good, I think the storage can be improved.

File: src/vaults/SingleStrategyWithdrawal.sol (hunk 1 above)
Comment (maruf0011) [reply src/vaults/SingleStrategyWithdrawal.sol:89]: For single strategy we are not locking any debt. each position is an epoch. So after closing a position we burn all the withdrawal debt.

File: src/vaults/SingleStrategyWithdrawal.sol [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+
+import {Vault} from "src/vaults/Vault.sol";
+
+contract SingleStrategyWithdrawalEscrow {
+    using SafeTransferLib for ERC20;
+    using FixedPointMathLib for uint256;
+
+    // Epoch counter
+    uint256 public currentEpoch;
```
[the remark this answers is not in the document]
Comment (maruf0011) [reply src/vaults/SingleStrategyWithdrawal.sol:15]: Possible, But Its better to keep as things as decoupled as possible.

[the remark this answers is not in the document]
Comment (maruf0011) [reply src/vaults/SingleStrategyWithdrawal.sol:15]: Also uint128 will keep both in same slot. But we are accessing `currentEpoch` a lot. and `lastResolvedUTCTime` only once. Will it make that much difference in gas optimization.

File: src/vaults/SingleStrategyWithdrawal.sol (hunk 1 above)
Comment (larryob) [reply src/vaults/SingleStrategyWithdrawal.sol]: Alright, looks good.

**Review by larryob — APPROVED** [review]
Thanks!

**Pull request merged**